File: Go/build_tree_from_preorder_and_inorder.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func build(pre []int, preStart int, preEnd int, inStart int, inPos map[int]int) *TreeNode {
    if preStart > preEnd {
        return nil
    }
    root := &TreeNode{Val: pre[preStart]}
    idx := inPos[root.Val]
    leftLen := idx - inStart

    root.Left = build(pre, preStart + 1, preStart + leftLen, inStart, inPos)
    root.Right = build(pre, preStart + leftLen + 1, preEnd, idx + 1, inPos)

    return root
}

func buildTree(pre []int, in []int) *TreeNode {
    inPos := make(map[int]int)
    for i := 0; i < len(in); i++ {
        inPos[in[i]] = i
    }
    return build(pre, 0, len(pre) - 1, 0, inPos)
}

File: Java/BuildTreeFromPreorderAndInorder.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    private TreeNode buildTree(
    int[] pre, int preStart, int preEnd,
    int inStart, Map<Integer, Integer> inPos) {

    if (preStart > preEnd) return null;
    TreeNode root = new TreeNode(pre[preStart]);
    int idx = inPos.get(root.val);
    int leftLen = idx - inStart;

    root.left = buildTree(pre, preStart + 1, preStart + leftLen, inStart, inPos);
    root.right = buildTree(pre, preStart + leftLen + 1, preEnd, idx + 1, inPos);

    return root;
  }

  // Time: O(n), Space: O(n)
  public TreeNode buildTree(int[] pre, int[] in) {
    Map<Integer, Integer> inPos = new HashMap<>();
    for (int i = 0; i < in.length; ++i)
      inPos.put(in[i], i);
    return buildTree(pre, 0, pre.length-1, 0, inPos);
  }
}